source/server/connection_handler_impl.cc

```diff
@@ -168,7 +168,17 @@ void ConnectionHandlerImpl::ActiveSocket::continueFilterChain(bool success) {
 
 void ConnectionHandlerImpl::ActiveListener::onAccept(
     Network::ConnectionSocketPtr&& socket, bool hand_off_restored_destination_connections) {
-  Network::Address::InstanceConstSharedPtr local_address = socket->localAddress();
+  // In case the connection was redirected transparently, set the local
+  // address as restored.
+  if (config_.transparent()) {
+    const Network::Address::InstanceConstSharedPtr local_address = socket->localAddress();
+    const Network::Address::InstanceConstSharedPtr bind_address = config_.socket().localAddress();
+    if (*local_address != *bind_address) {
```

Comment:
Nitpicking, but if the bind address is `0.0.0.0`, then this will always evaluate to true and we'll mark local address as restored even if connection was accepted on the listening port without any iptables magic.

Reply:
This is guarded by the `if (config_.transparent())` test above. In this case, the address should always be marked as restored. This is intentional.

Reply:
I don't think I have a way of telling whether an established connection is transparent.

Reply:
I disagree, IMHO connections accepted on listening port shouldn't be marked as restored (because they aren't), but that's fine with others, then I won't argue about it.

However, if that's the case, then the code is still wrong, since if the client connects to `127.0.0.1:port` then local address will be marked as restored if the listening address is `0.0.0.0:port`, but it won't be if the listening address is `127.0.0.1:port`.